sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs

```diff
@@ -36,34 +38,184 @@ public void Build(string? outputPath = null)
             outputPath ??= $".\\{GetType().Name}";
             outputPath = Path.GetFullPath(outputPath);
 
-            WriteBicepFile(this, outputPath);
+            Dictionary<Resource, List<Resource>> resourceTree = BuildResourceTree();
 
-            var queue = new Queue<IConstruct>();
-            queue.Enqueue(this);
+            var constructs = GetConstructs(true).ToArray();
+            foreach (var construct in constructs)
+            {
+                construct.GetResources(true);
+            }
+
+            ModuleConstruct? root = null;
+            BuildModuleConstructs(Root, resourceTree, null, ref root);
+
+            WriteBicepFile(root!, outputPath);
+
+            var queue = new Queue<ModuleConstruct>();
+            queue.Enqueue(root!);
             WriteConstructsByLevel(queue, outputPath);
         }
 
-        private void WriteConstructsByLevel(Queue<IConstruct> constructs, string outputPath)
+        private Dictionary<Resource, List<Resource>> BuildResourceTree()
+        {
+            var resources = GetResources(true).ToArray();
+            Dictionary<Resource, List<Resource>> resourceTree = new();
+            HashSet<Resource> visited = new();
+            foreach (var resource in resources)
+            {
+                VisitResource(resource, resourceTree, visited);
+            }
+
+            return resourceTree;
+        }
+
+        private void VisitResource(Resource resource, Dictionary<Resource, List<Resource>> resourceTree, HashSet<Resource> visited)
+        {
+            if (!visited.Add(resource))
+            {
+                return;
+            }
+
+            if (!resourceTree.ContainsKey(resource))
+            {
+                resourceTree[resource] = new List<Resource>();
+            }
+
+            if (resource.Parent != null)
+            {
+                if (!resourceTree.ContainsKey(resource.Parent))
+                {
+                    resourceTree[resource.Parent] = new List<Resource>();
+                }
+                resourceTree[resource.Parent].Add(resource);
+                VisitResource(resource.Parent, resourceTree, visited);
+            }
+        }
+
+        private void BuildModuleConstructs(Resource resource, Dictionary<Resource, List<Resource>> resourceTree, ModuleConstruct? parentScope, ref ModuleConstruct? root)
```

Comment:
Should this go into a separate class?  Don't have a strong opinion here but feels like a separate responsibility.

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/42035